android/greengrass-nucleus-android/src/main/java/com/aws/greengrass/android/managers/AndroidBaseComponentControl.java

```diff
@@ -0,0 +1,707 @@
+            if (receiverRegistered) {
+                Context context = contextProvider.getContext();
+                context.unregisterReceiver(receiver);
+            }
+            // Unconditionally release the semaphore regardless of the outcome
+            startAcknowledgmentSem.release();
+        }
+    }
+
+    private static Intent buildIntent(@NonNull Context context, @NonNull String packageName,
+                                      @NonNull String className, @NonNull String action,
+                                      @Nullable String[] arguments,
+                                      @Nullable Map<String, String> environment,
+                                      @NonNull Logger logger)
+            throws RuntimeException {
+
+        Intent intent = new Intent();
+        intent.setFlags(FLAG_ACTIVITY_NEW_TASK);
+        intent.setAction(action);
+        intent.setComponent(new ComponentName(packageName, className));
+        List<ResolveInfo> matches = context.getPackageManager().queryIntentServices(intent, 0);
+        // TODO: add service check
+        if (matches.size() == 1) {
+            // Add environmental variables
+            if (environment != null && !environment.isEmpty()) {
+                // Omit JAVA_HOME and HOME since they are not relevant for Android
+                environment.remove("JAVA_HOME");
+                environment.remove("HOME");
+
+                intent.putExtra(EXTRA_COMPONENT_ENVIRONMENT, (HashMap) environment);
+                intent.putExtra(EXTRA_MASTER_PACKAGE, context.getPackageName());
+                intent.putExtra(EXTRA_ARGUMENTS, arguments);
+            }
+        } else {
+            handleIntentResolutionError(matches, packageName, className, logger);
+        }
+        return intent;
+    }
+
+    private static void handleIntentResolutionError(List<ResolveInfo> matches, @NonNull String packageName,
+                                                    @NonNull String className, @NonNull Logger logger)
+            throws RuntimeException {
+        if (matches.size() == 0) {
+            logger.atError().kv(PACKAGE_NAME, packageName).kv(CLASS_NAME, className)
+                    .log("Service does not found");
+            throw new RuntimeException("Service with package " + packageName + " and class "
+                    + className + " couldn't found");
+        } else {
+            logger.atError().kv(PACKAGE_NAME, packageName).kv(CLASS_NAME, className)
+                    .log("Ambiguity in service");
+            throw new RuntimeException("Ambiguity in service with package " + packageName
+                    + " and class " + className + " found " + matches.size() + " matches");
+        }
+    }
+
+    private class IncomingHandler extends Handler {
+
+        IncomingHandler(Looper looper) {
+            super(looper);
+        }
+
+        @Override
+        public void handleMessage(Message msg) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 12. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.